File: 浙大数据结构/week11/散列2 Hashing.cpp
#include <stdio.h>

int isPrime(int x){
	if (x==1){
		return 0;
	}else if(x==2){
		return 1;
	}else if(x%2==0){
		return 0;
	}else{
		for(int i=2;i<x;i++){
			if(x%i==0){
				return 0;
			}
		}
		return 1;
	}
}

int main(){
	int m,n;
	scanf("%d %d",&m,&n);
	while(isPrime(m)==0){
		m++;
	}

//	//处理后的数据，初始化为-1 
//	int b[n]; 
//	for(int i=0;i<n;i++){
//		b[i]=-1;
//	}
	
	int c[m]={0};

	//读入数据 
	int a[n];
	for(int i=0;i<n;i++){
		scanf("%d",&a[i]);
	} 
	
	//开始处理数据
	for(int i=0;i<n;i++){
		int r=a[i]%m;
		//printf("%d %d\t",r,b[r]);
		if(c[r]==0){
			printf("%d",r);
			c[r]=1;
		}else{
			int tmp=r;
			int flag=0;
			for(int j=1;j<m;j++){
				tmp=(r+j*j)%m;
				if(c[tmp]==0){
					c[tmp]=1;
					flag=1;
					printf("%d",tmp);
					break;
				}
			}
			if(flag==0){
				printf("-");
			}
		}
		if(i<n-1){
			printf(" ");
		}
	}
	return 0;
}

File: 浙大数据结构/week11/散列2 Hashing.py
# -*- coding: utf-8 -*-
"""
Created on Mon May 21 20:49:19 2018

@author: Administrator
"""
import math

m,n=map(int,raw_input().strip().split())

num=map(int,raw_input().strip().split())
Hash={}

#找大于等于m的最小素数
def isPrime(x):
    if x==1:
        return False
    elif x==2:
        return True
    elif x%2==0:
        return False
    else:
        for i in range(3,int(math.sqrt(x)),2):
            if x%i==0:
                return False
        return True
    
while(True):
    if isPrime(m):
        break
    else:
        m+=1

ans=""       
#构造Hash表
for i in num:
    key=i%m
    if key in Hash.values():
        j=1
        flag=1
        while(key in Hash.values()):
            key+=j*j
            j+=1
            if(key>=m):
                flag=0
                break
        if(flag):
            Hash[i]=key
        else:
            Hash[i]='-'
    else:
        Hash[i]=key
    ans+=str(Hash[i])+" "
  
print(ans.strip())